src/Nethermind/Nethermind.Optimism/CL/P2P/OptimismCLP2P.cs

```diff
@@ -132,6 +175,64 @@ private async Task MainLoop(CancellationToken token)
         }
     }
 
+    private async Task<ExecutionPayloadV3?> TryRequestPayload
+        (ISession? session, ulong payloadNumber, Hash256 expectedHash, CancellationToken token)
+    {
+        try
+        {
+            CancellationTokenSource cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+            cancellationTokenSource.CancelAfter(TimeSpan.FromSeconds(5));
+            if (session is null) return null;
+            ExecutionPayloadV3? payload =
+                await session.DialAsync<PayloadByNumberProtocol, ulong, ExecutionPayloadV3?>(payloadNumber, cancellationTokenSource.Token);
+            if (payload is null || payload.BlockHash != expectedHash) return null;
+            return payload;
+        }
+        catch (Exception e)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Unable to request payload {payloadNumber} ({expectedHash}). Exception: {e}");
+            return null;
+        }
+    }
+
+    private async Task<ExecutionPayloadV3?> RequestPayload(ulong payloadNumber, Hash256 expectedHash, CancellationToken token)
+    {
+        if (_logger.IsInfo) _logger.Info($"Requesting missing payload. Number: {payloadNumber}, Expected hash: {expectedHash}");
+        try
+        {
+            ExecutionPayloadV3? response = null;
+            foreach (Multiaddress peer in _peerManager.GetPeers())
```

Comment:
Should we run this in parallel, that is, race `TryRequestPayload` for each peer and use the first response?

Reply:
I don't think it's good behavior